File: examples/python/campaignapp/service.py
import secrets
from typing import Protocol

from campaign.campaign import Campaign, CampaignSpec
from campaign.campaign_id import CampaignID
from campaign.short_link import ShortLink, ShortLinkSpec
from campaign.slug import Slug
from linkcampaign import (
    AddShortLinkRequest,
    AddShortLinkResponse,
    CreateCampaignRequest,
    CreateCampaignResponse,
    DeactivateShortLinkRequest,
    DeactivateShortLinkResponse,
    GetCampaignRequest,
    GetCampaignResponse,
    ShortLinkView,
)


class CampaignRepository(Protocol):

    def save(self, c: Campaign) -> None: ...

    def load(self, id: CampaignID) -> Campaign: ...


class CampaignService:

    def __init__(self, repo: CampaignRepository) -> None:
        self._repo = repo

    def create_campaign(self, req: CreateCampaignRequest) -> CreateCampaignResponse:
        spec = _to_campaign_spec(req)
        c = Campaign(spec)
        self._repo.save(c)
        return CreateCampaignResponse(
            campaign_id=str(c.id),
            name=str(c.name),
            links=_to_short_link_views(c.links),
        )

    def add_short_link(self, req: AddShortLinkRequest) -> AddShortLinkResponse:
        id = CampaignID(req.campaign_id)
        c = self._repo.load(id)
        c.add_short_link(
            ShortLinkSpec(slug=req.slug, target_url=req.target_url, active=True)
        )
        self._repo.save(c)
        return AddShortLinkResponse(
            campaign_id=str(c.id),
            links=_to_short_link_views(c.links),
        )

    def deactivate_short_link(
        self, req: DeactivateShortLinkRequest
    ) -> DeactivateShortLinkResponse:
        id = CampaignID(req.campaign_id)
        slug = Slug(req.slug)
        c = self._repo.load(id)
        c.deactivate_short_link(slug)
        self._repo.save(c)
        return DeactivateShortLinkResponse(
            campaign_id=str(c.id),
            links=_to_short_link_views(c.links),
        )

    def get_campaign(self, req: GetCampaignRequest) -> GetCampaignResponse:
        id = CampaignID(req.campaign_id)
        c = self._repo.load(id)
        return GetCampaignResponse(
            campaign_id=str(c.id),
            name=str(c.name),
            links=_to_short_link_views(c.links),
        )


def _to_campaign_spec(req: CreateCampaignRequest) -> CampaignSpec:
    links = tuple(
        ShortLinkSpec(slug=l.slug, target_url=l.target_url, active=True)
        for l in req.links
    )
    return CampaignSpec(id=_new_campaign_id_value(), name=req.name, links=links)


def _to_short_link_views(links: tuple[ShortLink, ...]) -> tuple[ShortLinkView, ...]:
    return tuple(
        ShortLinkView(slug=str(l.slug), target_url=str(l.target_url), active=l.active)
        for l in links
    )


def _new_campaign_id_value() -> str:
    return secrets.token_hex(8)

File: examples/running/campaignapp/service.go
package campaignapp

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"fmt"

	"github.com/verocorp/tesser-build/examples/running/campaign"
	"github.com/verocorp/tesser-build/examples/running/linkcampaign"
)

type CampaignRepository interface {
	Save(ctx context.Context, c campaign.Campaign) error
	Load(ctx context.Context, id campaign.CampaignID) (campaign.Campaign, error)
}

type CampaignService struct {
	repo CampaignRepository
}

func NewCampaignService(repo CampaignRepository) *CampaignService {
	return &CampaignService{repo: repo}
}

func (s *CampaignService) CreateCampaign(ctx context.Context, req linkcampaign.CreateCampaignRequest) (linkcampaign.CreateCampaignResponse, error) {
	spec := toCampaignSpec(req)

	c, err := campaign.NewCampaign(spec)
	if err != nil {
		return linkcampaign.CreateCampaignResponse{}, fmt.Errorf("invalid campaign: %w", err)
	}

	if err := s.repo.Save(ctx, c); err != nil {
		return linkcampaign.CreateCampaignResponse{}, fmt.Errorf("persist campaign %s: %w", c.ID(), err)
	}

	return linkcampaign.CreateCampaignResponse{
		CampaignID: c.ID().String(),
		Name:       c.Name().String(),
		Links:      toShortLinkViews(c.Links()),
	}, nil
}

func (s *CampaignService) AddShortLink(ctx context.Context, req linkcampaign.AddShortLinkRequest) (linkcampaign.AddShortLinkResponse, error) {
	id, err := campaign.NewCampaignID(req.CampaignID)
	if err != nil {
		return linkcampaign.AddShortLinkResponse{}, fmt.Errorf("invalid campaign id: %w", err)
	}

	c, err := s.repo.Load(ctx, id)
	if err != nil {
		return linkcampaign.AddShortLinkResponse{}, fmt.Errorf("load campaign %s: %w", req.CampaignID, err)
	}
	linkSpec := campaign.ShortLinkSpec{Slug: req.Slug, TargetURL: req.TargetURL}
	if err := c.AddShortLink(linkSpec); err != nil {
		return linkcampaign.AddShortLinkResponse{}, fmt.Errorf("add short link rejected: %w", err)
	}

	if err := s.repo.Save(ctx, c); err != nil {
		return linkcampaign.AddShortLinkResponse{}, fmt.Errorf("persist campaign %s: %w", req.CampaignID, err)
	}

	return linkcampaign.AddShortLinkResponse{
		CampaignID: c.ID().String(),
		Links:      toShortLinkViews(c.Links()),
	}, nil
}

func (s *CampaignService) DeactivateShortLink(ctx context.Context, req linkcampaign.DeactivateShortLinkRequest) (linkcampaign.DeactivateShortLinkResponse, error) {
	id, err := campaign.NewCampaignID(req.CampaignID)
	if err != nil {
		return linkcampaign.DeactivateShortLinkResponse{}, fmt.Errorf("invalid campaign id: %w", err)
	}
	slug, err := campaign.NewSlug(req.Slug)
	if err != nil {
		return linkcampaign.DeactivateShortLinkResponse{}, fmt.Errorf("invalid slug: %w", err)
	}

	c, err := s.repo.Load(ctx, id)
	if err != nil {
		return linkcampaign.DeactivateShortLinkResponse{}, fmt.Errorf("load campaign %s: %w", req.CampaignID, err)
	}
	if err := c.DeactivateShortLink(slug); err != nil {
		return linkcampaign.DeactivateShortLinkResponse{}, fmt.Errorf("deactivate short link rejected: %w", err)
	}

	if err := s.repo.Save(ctx, c); err != nil {
		return linkcampaign.DeactivateShortLinkResponse{}, fmt.Errorf("persist campaign %s: %w", req.CampaignID, err)
	}

	return linkcampaign.DeactivateShortLinkResponse{
		CampaignID: c.ID().String(),
		Links:      toShortLinkViews(c.Links()),
	}, nil
}

func (s *CampaignService) GetCampaign(ctx context.Context, req linkcampaign.GetCampaignRequest) (linkcampaign.GetCampaignResponse, error) {
	id, err := campaign.NewCampaignID(req.CampaignID)
	if err != nil {
		return linkcampaign.GetCampaignResponse{}, fmt.Errorf("invalid campaign id: %w", err)
	}

	c, err := s.repo.Load(ctx, id)
	if err != nil {
		return linkcampaign.GetCampaignResponse{}, fmt.Errorf("load campaign %s: %w", req.CampaignID, err)
	}

	return linkcampaign.GetCampaignResponse{
		CampaignID: c.ID().String(),
		Name:       c.Name().String(),
		Links:      toShortLinkViews(c.Links()),
	}, nil
}

func toCampaignSpec(req linkcampaign.CreateCampaignRequest) campaign.CampaignSpec {
	links := make([]campaign.ShortLinkSpec, len(req.Links))
	for i, l := range req.Links {
		links[i] = campaign.ShortLinkSpec{Slug: l.Slug, TargetURL: l.TargetURL, Active: true}
	}
	return campaign.CampaignSpec{
		ID:    newCampaignIDValue(),
		Name:  req.Name,
		Links: links,
	}
}

func toShortLinkViews(links []campaign.ShortLink) []linkcampaign.ShortLinkView {
	views := make([]linkcampaign.ShortLinkView, len(links))
	for i, l := range links {
		views[i] = linkcampaign.ShortLinkView{
			Slug:      l.Slug().String(),
			TargetURL: l.TargetURL().String(),
			Active:    l.Active(),
		}
	}
	return views
}

func newCampaignIDValue() string {
	b := make([]byte, 8)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}
